### ISSUE TITLE
[Feature request] Implement a shortcut/hotkey manager

### ISSUE THREAD

**Issue by schm1dtxbox** [RPCS3/rpcs3#5681]
In CS:GO(NPUB30589), ESC is used to enter the pause menu and exit menus when a keyboard and mouse is connected to a PS3, or when you are using the Basic Keyboard handler in rpcs3. Issue is, ESC is also used to exit fullscreen, so as a result, each time I need to enter the pause menu, I end up also exiting fullscreen, which is quite annoying. I would like it so that the Exit Fullscreen key can be rebinded in the rpcs3 options.

**Comment by vieiraa** [comment]
Would disabling ESC be a good solution? You can exit fullscreen with Alt+Enter as well.

**Comment by Megamouse** [comment]
I am working on a shortcut manager but it will probably take a long time to finish

**Comment by raelgc** [comment]
> Would disabling ESC be a good solution? You can exit fullscreen with Alt+Enter as well.

ESC would be great if:

- removed from exit fullscreen binding
- added to exit emulation (instead of exit fullscreen)

If someone can give me a quicktip (like told in which file this is defined), I can try change myself and open a PR.

**Comment by raelgc** [comment]
Should I just to into https://github.com/RPCS3/rpcs3/blob/master/rpcs3/rpcs3qt/gs_frame.cpp#L139 and change this:

```cpp
	case Qt::Key_Escape:
		if (visibility() == FullScreen) { toggle_fullscreen(); return; }
		break;
```

by this?

```cpp
	case Qt::Key_Escape:
		close();
		break;
```

**Comment by raelgc** [comment]
Ok, just tested this and the raw `close()` will not fire the proper event chain (check for quit confirmation, etc). 

`QWindow::close()` is the proper command. 

I edited the above comment to reflect this.

Fine if I open a PR?

**Comment by Megamouse** [comment]
I don't think people wanna close the game by clicking escape

**Comment by raelgc** [comment]
@Megamouse I'm an old guy, still used to press Escape to escape the current window :D

Now serious: I'm using a HTPC with Ubuntu and Retropie. Gamepad is my only device to interact with it (of course, I can attach a keyboard when required).

As most of users in the same situation, I've configured the gamepad home button to send the <kbd>Esc</kbd> to the emulators.

It works with most modern emulators: pcsx2, Dolphin, etc.

As rpcs3, they can be configured to exit the main program when emulation is finished. So, this way, I can get back to EmulationStation interface.

The esc key closing the emulation can solve this while you're working in the shortcut manager.

**Comment by schm1dtHL** [comment]
Just remembered that this issue existed, and I'd like to mention that you could have it so that holding escape would quit emulation, but not if you press it once, which would be ideal here until a shortcut manager is created imo.

**Comment by klepp0906** [comment]
was just coming to enquire about this.  a way to set up binds would be great.  whether it be for assigning toggling of fullscreen, stopping emulation, exiting rpcs3 alltogether etc.

that being said, im sure the developers are well enough aware and its on the proverbial laundry list : P

**Comment by lilcdj** [comment]
How's it going about this issue? I still can't use esc key to exit emulation even in --no-gui status.

**Comment by raelgc** [comment]
@lilcdj After talk with developers, I agree with them there is no need of this.

First: configure rpcs3 to exit after emulation is finished.

Then you can use gamepad -> keyboard mapper, like `xpadder` (on Windows) or `antimicro` (on Ubuntu) and configure your gamepad to, after you press the Home button, emit this sequence: `Esc` then `Ctrl+S`. It'll even produce a cool effect: exit from fullscreen then quit.

**Comment by klepp0906** [comment]
> I don't think people wanna close the game by clicking escape

what else would they want to close the game by?  every emulator ive ever used uses ESC to close.  Not that there is a problem with being different but..   so long as you have an toggleable option to confirm I cant imagine a more intuitive choice.  the fullscreen toggle is a bit more abstract in my opinion.  

> @lilcdj After talk with developers, I agree with them there is no need of this.
> 
> First: configure rpcs3 to exit after emulation is finished.
> 
> Then you can use gamepad -> keyboard mapper, like `xpadder` (on Windows) or `antimicro` (on Ubuntu) and configure your gamepad to, after you press the Home button, emit this sequence: `Esc` then `Ctrl+S`. It'll even produce a cool effect: exit from fullscreen then quit.

configuring rpcs3 to exit after emulation by default is problematic if you are playing at the pc and dont want to have to re-open RPCS3 and wait for the reload (if you have lots of titles etc) before starting another game.

As opposed to using ala frontend via no-gui where options within the frontends are usually present to circumvent the complication.  Launchbox for example allows small AHK scripts to accomplish this.

```
$Esc::
{
    Process, Close, {{{StartupEXE}}}
} 
```
Of course in some cases this will cause saves to not be written which could be a nightmare, and I havent tested with rpcs3. 

A shortcut/hotkey manager would be nice and is pretty par these days but beggers/choosers n all.   Just my 2c

**Comment by Nekotekina** [comment]
In theory, an option can be added to give Esc a second function (exit if not in fullscreen), although you'll have to push it twice in fullscreen. It sounds like a compromise with minimal effort to implement.

**Comment by klepp0906** [comment]
> In theory, an option can be added to give Esc a second function (exit if not in fullscreen), although you'll have to push it twice in fullscreen. It sounds like a compromise with minimal effort to implement.

I was actually thinking that very thing, just didnt suggest it as I didnt want to seem pushy.  You guys do so much its hard to make suggestions or argue points of view without feeling like youre coming across as ungrateful which I'd hate to do.

**Comment by dio-gh** [comment]
Modern applications never use Esc to exit completely. It would be a significant step back in UX to make such a hotkey the default, and is also completely unneeded.

I don't mind if hotkeys themselves become configurable however, I think RPCS3 definitely misses a field from it's settings menu like that; but I'm 100% against changing the default behaviour of Esc - or at least, of changing it to exiting the game.

An argument could be also made for making it completely unbounded by default, since Alt + Enter alone serves the purpose of going to and back from fullscreen both. I'd say it'll even be necessary, once there will be a native pause menu.

All in all though, the proper solution here seems to be adding a `Shortcuts` dialog to the GUI tab in the emulator settings, and make every currently hardcoded action rebindable. Toggles can't just keep increasing and cluttering the interface, and some of the default keybinds may not be comfortable, or even known to some - thinking about the screenshot hotkey here, for example. In my opinion, this would be the best for everyone.

**Comment by star69rem** [comment]
"Modern applications never use Esc to exit completely. It would be a significant step back in UX to make such a hotkey the default, and is also completely unneeded."

Is this satire? Virtually every emulator of note works this way OTHER than RPCS3.

Why do emulation authors seem laser focused on making sure that their programs are completely unusable?

**Comment by pupphelper** [comment]
Using any frontends or other launchers usually relay on "ESC" to close the emulator.  Perhaps esc could also exit only if the --nogui option is used.  Its a bit more code, but it would be even better if it was mappable in the options, which would give people the freedom to exit, fullscreen with the best key that suits their needs.  maybe leave alt-f4 alone but give the option for another exit key or another full-screen key.  key-combinations aren't the easiest to work with in other frontends.  Keep up the great work!

**Comment by hackmalife** [comment]
can this never be fixed? exiting with pressing ESC sucks so much...

**Comment by raelgc** [comment]
It makes sense to have <kbd>Esc</kbd> exiting when `--nogui` is `true`.

**Comment by elad335** [comment]
Perhaps, maybe, when savestates would be fast enough, then Esc can be bound to save a state then close the emulation so the user won't lose progress accidentally. Otherwise accidental clicks would result in a complete progress loss.

**Comment by Megamouse** [comment]
The shortcut/hotkey manager will be implemented as part of the Qt6 build, or shortly afterwards.
It is not worth the effort to add it in Qt5 due to various reasons.

**Comment by schm1dtxbox** [comment]
Looks like either PR 13186 or the shortcut manager should achieve what I wanted from the issue, since I just wanted to be able to unbind the key entirely to begin with for use in PS3 CSGO with Keyboard and Mouse, which is sweet.

**Comment by AniLeo** [comment]
Repurposed this ticket as a general shortcut manager following the discussion above

**Comment by raelgc** [comment]
Why not start with a simple `.ini` file with the shortcuts, then evolve to build the UX to reassign and save to that `.ini` file?

**Comment by Megamouse** [comment]
I already implemented most of this.
It's just not really compatible with Qt 5, so I just postponed it

**Comment by Darkhost1999** [comment]
The game window side is disabled and that must be what's being said
> It's just not really compatible with Qt 5

But these are configurable for main window and able to be edited under configuration -> shortcuts
![image](https://github.com/RPCS3/rpcs3/assets/60384196/52f89d50-84c3-4c25-90b7-5819b5dfdd88)

**Comment by raelgc** [comment]
This issue can be closed.

**Comment by Megamouse** [comment]
obviously not

**Comment by schm1dtxbox** [comment]
If the issue was still the original one I'd devised a while ago for unbinding Escape from exit fullscreen, I'd have probably closed it by now anyways ( since that request has been satisfied as of PR #13186 ), but clearly it's expanded in scope to focus on a more broad shortcut manager, so I see no reason to close it yet until that's completed in its entirety.

**Comment by oxyacetylene** [comment]
> obviously not

Agreed. I am used to using the physical "Pause" key on my keyboard to pause my emulator, because of years of playing games on Final Burn. I'd like the option to change it, if at all possible.

**Comment by SkyHighGam3r** [comment]
I'm surprised there isn't more discussion here regarding the general shortcut manager that this evolved into. The ability to change hotkeys to whatever a user's preferences are seems like something more people would be looking for. 

Personally I have a universal layout for these kinds of functions, and I apply it to every emulator that allows it. RPCS3 is one of the only ones I can't do that with unfortunately. 

(Just an example of some of what I use, to illustrate how much it differs from the defaults)

Close Program = Esc x 2
Screenshot = F10
Record = Insert
Savestate = F2
Loadstate = F4

**Comment by SilentException** [comment]
Testing new shortcut manager.
Official AppImage ﻿RPCS3 v0.0.33-16977-37d2317b Alpha | HEAD.
SteamOS Latest Stable 3.5.19.

Game window does not receive any keyboard input.
The main window shortcuts work fine and the log confirms it:

```
·! 0:00:11.848238 Shortcuts: main_window: Shortcut pressed: mw_exit_fullscreen (Esc)
·! 0:00:11.848313 GUI: Main window registered shortcut: mw_exit_fullscreen (Esc)
·! 0:02:46.001621 Shortcuts: main_window: Shortcut pressed: mw_stop (Ctrl+S)
·! 0:02:46.001643 GUI: Main window registered shortcut: mw_stop (Ctrl+S)
```

But in game window, pressing ESC (in SteamOS game mode using Steam+DPAD Left or in desktop mode using physical keyboard) I don't see anything happen neither on screen or in the log. Other shortcuts in game window don't seem to work either.

Two other discrepancies I've noticed on Linux AppImage:
Alt+F4 closes game window in desktop mode but not in game mode.
The shortcut to close main window is Ctrl+Q, reading previous issues it should be Alt+F4. In RPCS3 GUI in the File menu, the Exit menu item has shortcut Ctrl+Q. On windows there is no shortcut visible on Exit menu item.

**Comment by Megamouse** [comment]
This is not really a bug.
I forgot to update the docker image for the Linux builds.
It should work once we did that, which should happen sometime this week.
Currently the game window shortcuts won't work on AppImages since they are broken on older Qt versions.

**Comment by SilentException** [comment]
Noticed there was qt update. Testing on rpcs3-v0.0.33-16999-938306a7_linux64.AppImage.

```
·! 0:10:15.496259 Shortcuts: game_window: Shortcut pressed: gw_pause_play (PgUp)
·! 0:10:15.496278 GUI: Game window registered shortcut: gw_pause_play (PgUp)
```

Works fine but there is a misleading notification on screen "Press and hold the START button to resume".

```
·! 0:11:24.292088 Shortcuts: game_window: Shortcut pressed: gw_restart (F6)
·! 0:11:24.292110 GUI: Game window registered shortcut: gw_restart (F6)
```

Doesn't actually do anything. It does not restart emulation. Pressing mw_restart (in main window) works and restarts the emulated game.
BUT I later on figured that this shortcut actually only works if there is a save state. This was a bit confusing, having Restart shortcut both in main and game window but also serving different things. Maybe more user friendly way would be so that Restart shortcut restarts emulation and there is separate shortcut for loading save state in game window? Or at least renaming Restart to something different in game window since it doesn't actually do a restart?

Alt+F4 still doesn't work in game window or main window (tested only in game mode - maybe Wayland problem?).
Ctrl+Q works in main window but not in game window (also tested only in game mode).
Would be nice to have Alt+F4 working as it is kind of universal one for closing windows.

And thanks for working on this feature!

**Comment by Megamouse** [comment]
This has nothing to do with this pull request and no one will remember what you wrote :)

**Comment by SilentException** [comment]
> This has nothing to do with this pull request and no one will remember what you wrote :)

Alright. Just thought the test of shortcuts in issue about shortcuts might be helpful :)

**Comment by Megamouse** [comment]
"Works fine but there is a misleading notification on screen "Press and hold the START button to resume"."
- This is not really misleading. It's the default for exiting the Pause state and is unrelated to the shortcut and it works anyway.

"gw_restart"
- Someone changed how it works and used it for reloading savestates. No idea why.

"Alt+F4 still doesn't work in game window or main window"
- That's working fine on windows, so I'm assuming it's some issue with linux+Qt

**Comment by veke66** [comment]
To necro this, there is still no way of unbinding shortcuts in the manager, I had to open the config file and it just so happened that typing "none" removes the bind.
